addons/mail/models/mail_message.py

```diff
@@ -1133,6 +1138,16 @@ def _message_format_personalize(self, partner_id, messages_formatted=None, forma
                     'is_active': True,
                     'partner': {'id': partner_id, 'type': "partner"},
                 }
+            if not vals.get("needaction"):
+                # sudo: res.partner to format personalized message
+                partner = self.env['res.partner'].sudo().browse(partner_id)
```

Comment:
All notifications methods in mail are optimized, and here we are going back to a singleton-based check on partners, then users, without cache and anything ... especially we already computed all this data beforehand using optimized SQL (as we already know which recipients have internal users, ...), so I strongly advise rechecking the whole code and avoiding fetching data we already computed.

Reply:
Sort of propagating "recipients_data" (being in 15 in a big rebase don't have the 17.3 code here, but normally it is called that way).

Reply:
I understand the problem you are talking about. In the PR, https://github.com/odoo/odoo/pull/162065/files#diff-6039bdf622bf5417d796a812415303924107a9c8b37be05731aeaedd10beab21R1030 this has been removed and so that's why if we want to use `with_user()`, we have to specify the user one by one and it is indeed violating the optimization you had. What do you @seb-odoo think?